microbench/test_microbench.py

```diff
@@ -0,0 +1,129 @@
+"""
+These are not real tests, but microbenchmarks. The machinery to record the
+timing and display the results is inside conftest.py
+"""
+
+import pytest
+import time
+import _valgrind
+
+API_PARAMS = [
+    pytest.param('cpy', marks=pytest.mark.cpy),
+    pytest.param('hpy', marks=pytest.mark.hpy)
+    ]
+
+@pytest.fixture(params=API_PARAMS)
+def api(request):
+    return request.param
+
+@pytest.fixture
+def simple(request, api):
+    if api == 'cpy':
+        import cpy_simple
+        return cpy_simple
+    elif api == 'hpy':
+        import hpy_simple
+        return hpy_simple
+    else:
+        assert False, 'Unkown param: %s' % request.param
+
+@pytest.fixture
+def N(request):
+    n = 10000000
+    if request.config.option.fast:
+        n //= 100
+    if request.config.option.slow:
+        n *= 10
+    if _valgrind.lib.is_running_on_valgrind():
+        n //= 100
+    return n
+
+
+class TestModule:
+
+    def test_noargs(self, simple, timer, N):
+        with timer:
+            for i in range(N):
+                simple.noargs()
+
+    def test_onearg_None(self, simple, timer, N):
+        with timer:
+            for i in range(N):
+                simple.onearg(None)
+
+    def test_onearg_int(self, simple, timer, N):
+        with timer:
+            for i in range(N):
+                simple.onearg(i)
+
+    def test_varargs(self, simple, timer, N):
+        with timer:
+            for i in range(N):
+                simple.varargs(None, None)
+
+    def test_allocate_int(self, simple, timer, N):
+        with timer:
+            for i in range(N):
+                simple.allocate_int()
+
+    def test_allocate_tuple(self, api, simple, timer, N):
+        if api == 'hpy':
+            pytest.skip('Missing HPy_BuildValue')
+        with timer:
+            for i in range(N):
+                simple.allocate_tuple()
+
+
+class TestType:
+
+    def test_method_lookup(self, simple, timer, N):
+        obj = simple.Foo()
+        with timer:
+            for i in range(N):
+                # note: here we are NOT calling it, we want to measure just
+                # the lookup
+                obj.noargs
+
+    def test_method_lookup_ht(self, api, simple, timer, N):
+        obj = simple.HTFoo()
```

Comment:
Should we perhaps run all benchmarks against both `HTFoo` and `Foo`? Of course in the HPy case these are the same thing, but it would be good to compare with both static and heap types in the old C API.

Simplest option is to copy the tests and have `TestType` and `TestHeapType` . Other option would be to add another parameter to the test case.

Reply:
> Simplest option is to copy the tests and have TestType and TestHeapType . Other option would be to add another parameter to the test case.

or to substitute `simple` with a fixture `Foo` which yields both `simple. Foo` and `simple.HTFoo`.

Reply:
Further parameterizing fixtures is what I did with pytest-benchmark, but it requires rewriting the display logic when using the code in this PR, so I just added TestHeapType for now.